### DISCUSSION TITLE
Best model for coding: Wizard-Vicuna-13B-Uncensored-GPTQ?

### DISCUSSION THREAD

**Discussion by OWKenobi** [oobabooga/text-generation-webui#2036] [General]
what is in your opinion the best model for generating code, like python? 

model size is not a restriction. The best of course would be a link to a prequantized file, if you have any. Would be cool if in this thread we could collect the best ones out there with their pros and cons.

**Comment by m-spangenberg** [comment]


**Reply by OWKenobi** [reply]
That's an awesome answer, thank you so much! Can you verify which if the code snippets actually work? I often had the problem that A.I. was hallucinating good-looking code.

**Reply by m-spangenberg** [reply]
Hmm, definitely needs testing, you're right. My question was flawed anyway so I took down my comment. I don't want to confuse people or make them think certain models are better or worse than they really are because of error on my part. Let's hope for more specialized LLMs in the coding and tutoring space soon. :)

**Reply by Stoxis** [reply]
your response is blank m-spangenberg

**Comment by levicki** [comment]
All of them which responded by using OpenCV to convert an image to grayscale are on a level of a junior front-end developer importing LeftPad.

Try asking them this:

> Write a C# program using `System.Net.Http.HttpClient` class which downloads a file from a web server concurrently using 10 threads by specifying a `Range` HTTP request header so that each thread downloads different part of the file. Program should display total download progress as a percentage.

I'd be surprised if any of them respond with a code that works.

**Reply by OWKenobi** [reply]
I have no idea if this works, but this looks like a promising starting point, written by Wizard-Vicuna-13b-4bit. There is surely some flaw in this program, but just the way it loops over 10 files, to add up the downloaded bytes to actually display the download progress is surprisingly consistent!

    
    using System;
    using System.Collections.Generic;
    using System.IO;
    using System.Linq;
    using System.Net;
    using System.Threading;
    namespace DownloadFileConcurrentlyWithRangeHeader
    {
     class Program
     {
     static async Task Main(string[] args)
     {
     var httpClient = new HttpClient();
     await DownloadFileAsync("https://example.com/file", "filename.ext");
     Console.ReadLine();
     }
     static async Task DownloadFileAsync(string url, string fileName)
     {
     try
     {
     // Set up range header for each thread
     const int chunkSize = 5 \* 1024 \* 1024; // 5 MB per chunk
     var range = $"bytes={new byte[chunkSize].GetHashCode()}-";
     var client = new HttpClient();
     client.DefaultRequestHeaders.Add("Range", range);
     client.DefaultRequestHeaders.Accept.Clear();
     client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/octet-stream"));
     // Start downloading file concurrently
     var tasks = Enumerable.Repeat(client.GetStreamAsync($"{url}"), 10).ToArray();
     var files = await Task.WhenAll(tasks);
     foreach (var task in tasks)
     {
     var stream = await task;
     if (!await stream.CopyToAsync(System.IO.File.OpenWrite($@"C:\Temp\{fileName}")))
     {
     throw new Exception("Error writing file.");
     }
     }
     // Display total download progress as a percentage
     long totalBytesDownloaded = 0L;
     foreach (var file in files)
     {
     var bytesDownloaded = file.Length;
     totalBytesDownloaded += bytesDownloaded;
     Console.WriteLine($"Download complete: {bytesDownloaded}/{totalBytesDownloaded}%");
     }
     }
     catch (Exception ex)
     {
     Console.WriteLine(ex.Message);
     }
     }
     }
    }

**Comment by OWKenobi** [comment]
Ok, thanks to @m-spangenberg , here is my number 1 candidate (as of now):

### Wizard-Vicuna-13B-Uncensored-GPTQ

    Link:  https://huggingface.co/TheBloke/Wizard-Vicuna-13B-Uncensored-GPTQ/tree/main
    Launch parameters: --wbits 4 --groupsize 128 --model_type llama

Pros: 
         - can code decently (still full of errors, but it develops a clear idea).
          - can do consistent roleplay with a single line as instruction, and even impersonate different actors.
            _Example: You will impersonate the computer of star ship enterprise. I am the Captain. This is a roleplay_
         - can do some simple math correctly:
           _Example: Evaluate step by step how much is 4 * 5 * 5, explain it in simple words._
        - only 9GB of VRAM needed
        -  seems not to answer itself (didn't happen a single time during my testing)
        -  uncensored, so no "As an A.I. model..." responses!

Cons: 
       - still hallucinates, doesn't know when it does so.

**Reply by levicki** [reply]
@OWKenobi 
> - uncensored, so no "As an A.I. model..." responses!

While it may be uncensored and not refuse to answer it is still very biased on a range of controversial topics.

**Comment by levicki** [comment]
@OWKenobi Eh, where do I begin...

> `const int chunkSize = 5 * 1024 * 1024; // 5 MB per chunk`

Wrong. Chunk size has to be calculated based on file size and number of threads and the last chunk size may not be the same as the previous ones.

For a human it would be obvious that some structure is needed to store chunk offsets and lengths, but to the AI with the reasoning of 3 year old? Apparently not.

> `var range = $"bytes={new byte[chunkSize].GetHashCode()}-";`

Wrong on so many levels...

1. `new byte[chunkSize]` allocates a 5 MB byte array but doesn't store it in any variable.
2. The range string value is always in the form of `bytes=N-` which means every thread would download from a set point to the end of file. That would not only waste bandwidth but also fail when they try to write over each other's part of output file.
3. What's that `GetHashCode()` for anyway? It apparently couldn't do the math needed to split the file to 10 parts based on the file size so it just hallucinated something random instead. The hash returned will be different for every new byte array constructed and there is high probability it will be a number which is out of available range of bytes for download.

> `var client = new HttpClient();`

Wrong again, doesn't specify neither that it needs automatic decompression for GZip and deflate server responses, nor max connections per server (default is 1 so it wouldn't download in parallel even if the code wasn't broken to begin with).

> `if (!await stream.CopyToAsync(System.IO.File.OpenWrite($@"C:\Temp\{fileName}")))`

Wrong yet again, from the documentation for `OpenWrite` method (emphasis mine):

> The [OpenWrite](https://learn.microsoft.com/en-us/dotnet/api/system.io.file.openwrite?view=net-7.0) method opens a file if one already exists for the file path, or creates a new file if one does not exist. For an existing file, it does not append the new text to the existing text. **Instead, it overwrites the existing characters with the new characters.**

So it failed to understand it needs offsets and seeking, but there's more:

> This method is equivalent to the [FileStream(String, FileMode, FileAccess, FileShare)](https://learn.microsoft.com/en-us/dotnet/api/system.io.filestream.-ctor?view=net-7.0#system-io-filestream-ctor(system-string-system-io-filemode-system-io-fileaccess-system-io-fileshare)) constructor overload with file mode set to [OpenOrCreate](https://learn.microsoft.com/en-us/dotnet/api/system.io.filemode?view=net-7.0#system-io-filemode-openorcreate), the access set to [Write](https://learn.microsoft.com/en-us/dotnet/api/system.io.fileaccess?view=net-7.0#system-io-fileaccess-write), and the share mode set to [None](https://learn.microsoft.com/en-us/dotnet/api/system.io.fileshare?view=net-7.0#system-io-fileshare-none).

It also failed to understand that it needs to specify share mode other than `None` if it wants to write to the same file from 10 threads at once.

I won't even comment on progress calculation, it's pointless because it needs to happen during downloading of each chunk and to update a single variable from multiple threads.

If this is the level of coding skill of current AI models (and I know junior developers who already rely on ChatGPT to help them code), then God help us all unless the software companies ban use of it until it can actually produce something sensible.

Don't get me wrong, AI is a great tool and it could help with menial tasks and boilerplate code, but in the time it took me to check this garbage I could have written it myself.

The main issue I see with its use in its current form is that a junior developer asking AI to write code for them won't have the means (read: necessary knowledge and experience) to validate AI code output other than "it compiles so it could be ok".

Finally, this example was requesting use of commonly used classes like `HttpClient` (StackOverflow is full of it so it was porobably in the training data). Trust me, you don't want to know what garbage comes out when you ask for more complex things. Or maybe you do?

Here, try this one:
> Write a C function with the signature HRESULT SignFile(LPWCSTR FilePath, PCERT_CONTEXT pSignerCert) which will digitally sign a PE file using SignerSign() API from mssign32.dll.

**Reply by valden80** [reply]
Well, it's should be expected. Right now, this is not actual "Artificial Intelligence", it's just "Neural Network". 
It's don't really understand what is it generate, so for now it's just dumb to expect a serious working code from it. (More serious then hello world or typical simple popular tasks, like a sorting of array or factorial calculation).
Current NN's don't have logical block, they don't actually resolve problems/tasks. If you asking them to do some math - they don't calc. They's look in their "memory" - that come from dataset, and "pull out" something similar and relative to the question. Like a scholar who don't know the answer (and can't do math at all), so  just looking for it from a neighbors in the class.

**Reply by levicki** [reply]
@valden80 
> for now it's just dumb to expect a serious working code from it.

Seems it went over your head that I wasn't expecting the correct answer &mdash; I was merely pointing out to the OP and anyone responding to or reading this discussion that the code generated by the AI is unreliable and unless they have enough knowledge to validate its output they will be taking a great risk by incorporating it in any project. At present, that risk is far greater than picking up a random bit of code from Stack Overflow.

> More serious then hello world or typical simple popular tasks, like a sorting of array or factorial calculation

Even for those you are better off using Stack Overflow.

> Like a scholar who don't know the answer (and can't do math at all), so just looking for it from a neighbors in the class.

Not really &mdash; it actually works by predicting the next likely token.

For example:

**You:** Who painted Mona Lisa?
**Assistant:** Leonardo Da Vinci

It doesn't know the answer &mdash; it breaks down the question into a bunch of tokens and predicts the set of tokens that are most likely to follow them which so happens to be the answer because the tokens representing the answer were close to the tokens representing the question in the training dataset.

In other words, there's no intelligence in it at all. It should be more fair to call it NTP (as in Next Token Predictor).

It's but a [stochastic parrot](https://dl.acm.org/doi/10.1145/3442188.3445922) which can mimic human reasoning to some (extremely limited) extent.

**Reply by OWKenobi** [reply]
Yes, that is the problem; it just makes too many mistakes. But at least it has an idea, a direction, knows some toolchains that it thinks it should use. I hope with bigger models (or more training) the mistakes get fewer. And it actually helps me; I never coded C#, so I wouldn't even know how to catch an Exception or how to load a dependency, or which dependencies to load; so for me it would still save some time, even with all these errors inside.

Meanwhile I have an idea: what if we told it to just produce pseudo-code? It will still help, but it would not feel pushed to produce content it doesn't yet understand fully.

**Reply by levicki** [reply]
I am afraid that larger models won't help with accuracy or the ability to understand more complex requests.

Take the second example I gave for a C program. The task sounds deceptively simple &mdash; I even told it which function to use (`SignerSign`), but in order to use it, it would have to know that it has to load `mssign32.dll` dynamically using `LoadLibrary` and to get the function pointer using `GetProcAddress` since there is no import library provided for `mssign32.dll` and its functions cannot be called directly.

Furthermore, all C structures needed to work with that function are [documented online](https://learn.microsoft.com/en-us/windows/win32/seccrypto/signersign), but in order to use them it would have to have the ability to find and "read" this documentation and then dive recursively into the C structure definitions provided to create an appropriate header file so it can actually call the function with required parameters, not to mention it would need the ability to "fill in the blanks" (i.e. deduce proper values for other `SignerSign` parameters based on stated intent).

Instead of telling it to produce code, you can ask it to outline the steps needed to accomplish the task, but if it doesn't understand the task and doesn't have the domain-specific knowledge it will give you the wrong steps too.

For example, asking ChatGPT to write a C# program to sign a PE executable will result in it first trying to use a commercial C# library that can accomplish that task which is pretty good. However, if you tell it not to use any commercial code, then ChatGPT will spit out this "solution":
```
using System;
using System.IO;
using System.Security.Cryptography;
using System.Security.Cryptography.Pkcs;
using System.Security.Cryptography.X509Certificates;

class Program
{
    static void Main(string[] args)
    {
        // Load the code signing certificate
        X509Certificate2 certificate = new X509Certificate2("path/to/certificate.pfx", "certificate_password");

        // Load the PE executable file
        byte[] fileBytes = File.ReadAllBytes("path/to/PE/executable");

        // Create a signed manifest based on the file
        byte[] signedBytes = CreateSignedManifest(fileBytes, certificate);

        // Save the signed file
        File.WriteAllBytes("path/to/signed/PE/executable", signedBytes);

        Console.WriteLine("File signed successfully.");
    }

    static byte[] CreateSignedManifest(byte[] fileBytes, X509Certificate2 certificate)
    {
        // Create a ContentInfo object to represent the file data
        ContentInfo content = new ContentInfo(fileBytes);

        // Create a SignedCms object to represent the signed data
        SignedCms signedData = new SignedCms(content);

        // Create a signer information object
        CmsSigner signer = new CmsSigner(certificate);

        // Add the signer information to the signed data object
        signedData.ComputeSignature(signer);

        // Encode the signed data object to create the signed manifest
        byte[] signedBytes = signedData.Encode();

        return signedBytes;
    }
}
```
Although the code appears fully functional (I can't spot any errors that would cause the compilation to fail) if you have the domain-specific knowledge about PE file format you can immediately spot that it has absolutely no understanding of what signing a PE executable involves &mdash; you don't sign a PE file the same way you sign an encrypted message.

PE file has a well defined structure and its signature is stored in a SecurityDirectory which means you have to parse the file structure, produce hashes, produce the signature, and then append the signature and the certificate chain by expanding the file to contain them and making appropriate changes in the file header to point to the added security data.

If it doesn't understand all that (and currently it doesn't), then it will give you wrong steps to follow if you ask for instructions or pseudo-code, and unless you know it's wrong you will end up following wrong instructions just like you would end up trying to run this code that does the wrong thing.

> And it actually helps me; I never coded C#, so I wouldn't even know how to catch an Exception or how to load a dependency, or which dependencies to load; so for me it would still save some time, even with all these errors inside.

I don't think that learning how to do those things from broken code generated by AI is a good idea. You would be better served by just googling "C# exceptions" and reading the properly documented examples.
